docs/StardustDocs/topics/readSqlFromCustomDatabase.md

```diff
@@ -0,0 +1,167 @@
+public object HSQLDB : DbType("hsqldb") {
+    override val driverClassName: String
+        get() = "org.hsqldb.jdbcDriver"
+
+    override fun convertSqlTypeToColumnSchemaValue(tableColumnMetadata: TableColumnMetadata): ColumnSchema? {
+        return null
+    }
+
+    override fun isSystemTable(tableMetadata: TableMetadata): Boolean {
+        val locale = Locale.getDefault()
+        fun String?.containsWithLowercase(substr: String) = this?.lowercase(locale)?.contains(substr) == true
+        val schemaName = tableMetadata.schemaName
+        val name = tableMetadata.name
+        return schemaName.containsWithLowercase("information_schema") ||
+            schemaName.containsWithLowercase("system") ||
+            name.containsWithLowercase("system_")
+    }
+
+    override fun buildTableMetadata(tables: ResultSet): TableMetadata =
+        TableMetadata(
+            tables.getString("TABLE_NAME"),
+            tables.getString("TABLE_SCHEM"),
+            tables.getString("TABLE_CAT"),
+        )
+
+    override fun convertSqlTypeToKType(tableColumnMetadata: TableColumnMetadata): KType? {
+        return null
+    }
+}
+```
+
+**Defining Helper Functions**
+
+Define utility functions to manage database connections and tables.
+
+```kotlin
+import java.sql.Connection
+
+const val URL = "jdbc:hsqldb:hsql://localhost/testdb"
+const val USER_NAME = "SA"
+const val PASSWORD = ""
+
+
+fun removeTable(con: Connection): Int {
+    val stmt = con.createStatement()
+    return stmt.executeUpdate("""DROP TABLE orders""")
+}
+
+fun createAndPopulateTable(con: Connection) {
+    var stmt = con.createStatement()
+    stmt.executeUpdate(
+        """CREATE TABLE IF NOT EXISTS orders (
+                        id INT GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
+                        item VARCHAR(50) NOT NULL,
+                        price DOUBLE NOT NULL,
+                        order_date DATE
+                    );
+                """.trimIndent()
+    )
+
+    stmt.executeUpdate(
+        """INSERT INTO orders (item, price, order_date) 
+                        VALUES ('Laptop', 1500.00, NOW())""".trimIndent()
+    )
+
+    stmt.executeUpdate(
+        """INSERT INTO orders (item, price, order_date) 
+                        VALUES ('Smartphone', 700.00, NOW())""".trimIndent()
+    )
+}
+```
+
+**Define the Table Schema**
+
+Use the `@DataSchema` annotation to define a [**custom data schema**](schemasCustom.md) for the `orders` table.
+
+```kotlin
+@DataSchema
+interface Orders {
+    val id: Int
+    val item: String
+    val price: Double
+    val orderDate: java.util.Date
+}
+```
+
+**End-to-End Example**
+
+Finally, use the following code to create, populate, read, and delete the table in HSQLDB.
+
+```kotlin
+fun main() {
+    DriverManager.getConnection(URL, USER_NAME, PASSWORD).use { con ->
+        createAndPopulateTable(con)
+
+        val df = con.readDataFrame("SELECT * FROM orders", dbType = HSQLDB).rename {
+            all()
```

Comment:
If you can, try to fit `rename` and `into` on the same line. That makes DataFrame operations easier to read and understand at a glance.